pymc3/tests/test_sampling.py

```diff
@@ -121,6 +121,37 @@ def test_iter_sample(self):
             for i, trace in enumerate(samps):
                 assert i == len(trace) - 1, "Trace does not have correct length."
 
+    def test_sample_does_not_modify_start_as_list_of_dicts(self):
+        # make sure pm.sample does not modify the 'start_list' passed as an argument
+        # see https://github.com/pymc-devs/pymc3/pull/4458
+        start_list = [{"mu1": 10}, {"mu2": 15}]
+        with self.model:
+            mu1 = pm.Normal("mu1", mu=0, sd=5)
+            mu2 = pm.Normal("mu2", mu=0, sd=1)
+            trace = pm.sample(
+                step=pm.Metropolis(),
+                tune=5,
+                draws=10,
+                chains=2,
+                start=start_list,
+            )
+        assert start_list == [{"mu1": 10}, {"mu2": 15}]
+
+    def test_sample_does_not_modify_start_as_dict(self):
+        # make sure pm.sample does not modify the 'start_dict' passed as an argument.
+        # see https://github.com/pymc-devs/pymc3/pull/4458
+        start_dict = {"X0_mu": 25}
+        with self.model:
+            X0_mu = pm.Lognormal("X0_mu", mu=np.log(0.25), sd=0.10)
+            trace = pm.sample(
+                step=pm.Metropolis(),
+                tune=5,
+                draws=10,
+                chains=3,
+                start=start_dict,
+            )
+        assert start_dict == {"X0_mu": 25}
+
```

Comment:
Let's not overcomplicate the tests. Also we should test both `pm.sample` and `pm.find_MAP`.
* no need to reuse the complicated `self.model` model
* variable names and parameters don't matter
* distribution should be transformed by default (Uniform, Lognormal, ...)
* everything in the same test case so the compilation is done just once
```
with pm.Model():
    pm.Lognormal("untransformed")

    # test that find_MAP doesn't change the start dict
    start = {"untransformed": 2}
    pm.find_MAP(start=start, niter=5)
    assert start == {"untransformed": 2}
    
    # check that sample doesn't change it either
    start = {"untransformed": 0.5}
    ...

    # and also not if start is different for each chain
    start = [{"untransformed": 2}, {"untransformed": 0.5}]
    ...
```

Reply:
I did remove these lines but It came back from nowhere (I might have fetched before committing).
Anyway, this time I will make sure that these lines does not come back.

Reply:
It would be better if we take this test outside from the `TestSample` class.